### PR TITLE
Allow value converters to convert nulls

### LINKED ISSUE #13850: Allow HasConversion/ValueConverters to convert nulls
When using `FromSql` to create a custom filtered data set, any column conversions defined on the `DbQuery` object don't get applied.

I think the column conversions should be included automatically around the sub-query supplied to `FromSql`, similar to how a sub-query is used if more predicates are added to the `IQueryable`.

```c#
    public class Model
    {
        public int Id { get; set; }
        public bool Flag { get; set; }
    }

    public partial class MyDbContext : DbContext
    {
        public DbQuery<Model> Models { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
                    modelBuilder.Query<Models>(
                        entity =>
                        {
                            entity.ToView("dbo.Models");
                            // Convert null Flag to false on way in
                            entity.Property(e => e.Flag).HasConversion<bool?>(f => f, t => t ?? false);
                        });
        }
    }

    // ...

    var models = dbContext.Models.FromSql("SELECT * FROM dbo.Models").ToList();
```

This, I think, should result in SQL similar to this being executed...
```SQL
SELECT [e].[Id], COALESCE([e].[Flag], 0) AS [Flag]
FROM (
    SELECT * FROM dbo.Models) AS [e];
```
But, in reality, the plain SQL from `FromSql` is executed
```SQL
SELECT * FROM dbo.Models
```

EF Core version: 2.1.4
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 7
IDE: Visual Studio 2017 15.8.5

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #13850

### CODE REVIEW

**Comment by ajcvickers** [comment]
@dotnet/efteam Could someone review this today so that I can have it merged ready for the the live-stream PR?

**Comment by ajcvickers** [comment]
@dotnet/efteam Change what I am going to show in the stream, so no rush here again.

File: src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs [hunk 1]
```diff
@@ -701,7 +701,30 @@ public static void SetIsStored(this IMutableProperty property, bool? value)
...
+            property.TryGetDefaultValue(out var defaultValue);
+            return defaultValue;
+        }
+
+        /// <summary>
+        ///     Returns the object that is used as the default value for the column this property is mapped to.
+        /// </summary>
+        /// <param name="property"> The property. </param>
+        /// <param name="defaultValue"> The default value, or the CLR default if no explicit default has been set. </param>
+        /// <returns> The object that is used as the default value for the column this property is mapped to. </returns>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs:714]: Update the `<returns>` text

File: src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs [hunk 2]
```diff
@@ -710,20 +733,38 @@ public static void SetIsStored(this IMutableProperty property, bool? value)
...
+            return defaultValue;
+        }
+
+        /// <summary>
+        ///     Returns the object that is used as the default value for the column this property is mapped to.
+        /// </summary>
+        /// <param name="property"> The property. </param>
+        /// <param name="storeObject"> The identifier of the table-like store object containing the column. </param>
+        /// <param name="defaultValue"> The default value, or the CLR default if no explicit default has been set. </param>
+        /// <returns> The object that is used as the default value for the column this property is mapped to. </returns>
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalPropertyExtensions.cs:747]: and here

File: src/EFCore/Storage/ValueConversion/BoolToZeroOneConverter.cs [hunk 3]
```diff
@@ -30,14 +30,38 @@ public BoolToZeroOneConverter(ConverterMappingHints? mappingHints = null)
...
+                                        ? (ulong)0
+                                        : type == typeof(byte)
+                                            ? (byte)0
+                                            : type == typeof(decimal)
+                                                ? (decimal)0
+                                                : type == typeof(double)
+                                                    ? (double)0
+                                                    : type == typeof(float)
+                                                        ? (float)0
+                                                        : (object)0);
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/BoolToZeroOneConverter.cs:64]: 😮

File: src/EFCore/Storage/ValueConversion/BytesToStringConverter.cs [hunk 4]
```diff
@@ -20,9 +20,9 @@ public class BytesToStringConverter : ValueConverter<byte[], string>
         public BytesToStringConverter(
             ConverterMappingHints? mappingHints = null)
             : base(
-                // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
                 v => v == null ? null! : Convert.ToBase64String(v),
                 v => v == null ? null! : Convert.FromBase64String(v),
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/BytesToStringConverter.cs:24]:
```suggestion
                v => v == null ? null : Convert.ToBase64String(v),
                v => v == null ? null : Convert.FromBase64String(v),
```

File: src/EFCore/Storage/ValueConversion/IPAddressToBytesConverter.cs [hunk 5]
```diff
@@ -21,9 +21,9 @@ public class IPAddressToBytesConverter : ValueConverter<IPAddress, byte[]>
         /// </param>
         public IPAddressToBytesConverter(ConverterMappingHints? mappingHints = null)
             : base(
-                // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
                 v => v == null ? default! : v.GetAddressBytes(),
                 v => v == null ? default! : new IPAddress(v),
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/IPAddressToBytesConverter.cs:25]:
```suggestion
                v => v == null ? default : v.GetAddressBytes(),
                v => v == null ? default : new IPAddress(v),
```

File: src/EFCore/Storage/ValueConversion/IPAddressToStringConverter.cs [hunk 6]
```diff
@@ -36,12 +37,10 @@ public IPAddressToStringConverter(ConverterMappingHints? mappingHints = null)
         public static ValueConverterInfo DefaultInfo { get; }
             = new(typeof(IPAddress), typeof(string), i => new IPAddressToStringConverter(i.MappingHints), _defaultHints);
 
-        private static new Expression<Func<IPAddress, string>> ToString()
-            // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
+        private static new Expression<Func<IPAddress?, string?>> ToString()
             => v => v == null ? default! : v.ToString();
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/IPAddressToStringConverter.cs:41]:
```suggestion
            => v => v == null ? default : v.ToString();
```

File: src/EFCore/Storage/ValueConversion/IPAddressToStringConverter.cs [hunk 7]
```diff
@@ -36,12 +37,10 @@ public IPAddressToStringConverter(ConverterMappingHints? mappingHints = null)
...
 
-        private static Expression<Func<string, IPAddress>> ToIPAddress()
-            // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
+        private static Expression<Func<string?, IPAddress?>> ToIPAddress()
             => v => v == null ? default! : IPAddress.Parse(v);
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/IPAddressToStringConverter.cs:44]:
```suggestion
            => v => v == null ? default : IPAddress.Parse(v);
```

File: src/EFCore/Storage/ValueConversion/PhysicalAddressToBytesConverter.cs [hunk 8]
```diff
@@ -21,9 +21,9 @@ public class PhysicalAddressToBytesConverter : ValueConverter<PhysicalAddress, b
         /// </param>
         public PhysicalAddressToBytesConverter(ConverterMappingHints? mappingHints = null)
             : base(
-                // TODO-NULLABLE: Null is already sanitized externally, clean up as part of #13850
                 v => v == null ? default! : v.GetAddressBytes(),
                 v => v == null ? default! : new PhysicalAddress(v),
```
Comment (AndriySvyryd) [inline src/EFCore/Storage/ValueConversion/PhysicalAddressToBytesConverter.cs:25]:
```suggestion
                v => v == null ? default : v.GetAddressBytes(),
                v => v == null ? default : new PhysicalAddress(v),
```

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**